### PR TITLE
Enable VP9 HW Encoder in msdk omx plugin

### PR DESCRIPTION
**Pull request by TianmiChen**
Signed-off-by: tianmi.chen <tianmi.chen@intel.com>

### CODE REVIEW

File: omx_components/src/mfx_omx_ports.cpp [hunk 1]
```diff
@@ -163,6 +163,10 @@ bool mfx_omx_is_index_valid(OMX_INDEXTYPE index, MfxOmxPortId port_id)
             if (MfxOmx_IndexIntelDisableDeblockingIdc == index) return true;
             break;
         case MfxOmxPortVideo_vp9ve:
+            if (OMX_IndexParamVideoVp9 == index) return true;
+            if (OMX_IndexParamVideoBitrate == index) return true;
+            if (OMX_IndexParamVideoAndroidVp8Encoder == index) return true;
```
Comment (DenWolf) [inline omx_components/src/mfx_omx_ports.cpp]: Why vp8 encoder is presented here - "OMX_IndexParamVideoAndroid**Vp8**Encoder " ?

Comment (TianmiChen) [reply omx_components/src/mfx_omx_ports.cpp]: Hi Denis,
Yes, it was a little confused for me before. I checked AOSP code, this index is shared by both VP8 and VP9 encoder in android framework.  Please check below structure for more details.

/**
 * Android specific VP8/VP9 encoder params
 *
 * STRUCT MEMBERS:
 *  nSize                      : Size of the structure in bytes
 *  nVersion                   : OMX specification version information
 *  nPortIndex                 : Port that this structure applies to
 *  nKeyFrameInterval          : Key frame interval in frames
 *  eTemporalPattern           : Type of temporal layer pattern
 *  nTemporalLayerCount        : Number of temporal coding layers
 *  nTemporalLayerBitrateRatio : Bitrate ratio allocation between temporal
 *                               streams in percentage
 *  nMinQuantizer              : Minimum (best quality) quantizer
 *  nMaxQuantizer              : Maximum (worst quality) quantizer
 */
typedef struct OMX_VIDEO_PARAM_ANDROID_VP8ENCODERTYPE {
    OMX_U32 nSize;
    OMX_VERSIONTYPE nVersion;
    OMX_U32 nPortIndex;
    OMX_U32 nKeyFrameInterval;        // distance between consecutive key_frames (including one
                                      // of the key_frames). 0 means interval is unspecified and
                                      // can be freely chosen by the codec. 1 means a stream of
                                      // only key_frames.

    OMX_VIDEO_ANDROID_VPXTEMPORALLAYERPATTERNTYPE eTemporalPattern;
    OMX_U32 nTemporalLayerCount;
    OMX_U32 nTemporalLayerBitrateRatio[OMX_VIDEO_ANDROID_MAXVP8TEMPORALLAYERS];
    OMX_U32 nMinQuantizer;
    OMX_U32 nMaxQuantizer;
} OMX_VIDEO_PARAM_ANDROID_VP8ENCODERTYPE;

**Comment by DenWolf** [comment]
@Losev-Ivan , could you please also take a look to the patch?

File: omx_components/src/mfx_omx_ports.cpp (hunk 1 above)
Comment (Losev-Ivan) [reply omx_components/src/mfx_omx_ports.cpp]: Yes, according to code some parameters for vp9e we may get only from OMX_IndexParamVideoAndroidVp8Encoder, therefore I think it's correct.

Comment (TianmiChen) [reply omx_components/src/mfx_omx_ports.cpp]: Thanks lvan. So is it ok to merge the PR now?

File: omx_utils/include/mfx_omx_types.h [hunk 2]
```diff
@@ -341,6 +341,10 @@ struct MfxOmxParamsWrapper: public T
             ext_buf[idx].roi.Header.BufferId = MFX_EXTBUFF_ENCODER_ROI;
             ext_buf[idx].roi.Header.BufferSz = sizeof(mfxExtEncoderROI);
             return idx;
+          case MFX_EXTBUFF_VP9_PARAM:
+            ext_buf[idx].roi.Header.BufferId = MFX_EXTBUFF_VP9_PARAM;
```
Comment (Losev-Ivan) [inline omx_utils/include/mfx_omx_types.h]: I'm not sure that this is correct, @DenWolf , can you look at this?

File: omx_utils/src/mfx_omx_defaults.cpp [hunk 3]
```diff
@@ -115,6 +115,18 @@ void mfx_omx_set_defaults_mfxVideoParam_enc(mfxVideoParam* params)
...
         params->mfx.NumRefFrame = 1;
         break;
+	case MFX_CODEC_VP9:
+        params->mfx.CodecProfile = MFX_PROFILE_VP9_0;
+        params->mfx.CodecLevel = 0;
+        params->mfx.TargetUsage = MFX_TARGETUSAGE_BEST_SPEED;
+        params->mfx.FrameInfo.PicStruct = MFX_PICSTRUCT_PROGRESSIVE;
+        params->mfx.TargetKbps = 3000;
+        params->mfx.RateControlMethod = MFX_RATECONTROL_CBR;
+        params->mfx.GopPicSize = 16;
```
Comment (DenWolf) [inline omx_utils/src/mfx_omx_defaults.cpp]: Minor comment - I suggest to set GopPicSize=15 (or 30 aka 30fps) for VP9

File: omx_utils/include/mfx_omx_types.h (hunk 2 above)
Comment (DenWolf) [reply omx_utils/include/mfx_omx_types.h]: Agree, I don't think that can set ROI for vp9, so I suggest to remove this part

File: omx_utils/src/mfx_omx_defaults.cpp (hunk 3 above)
Comment (TianmiChen) [reply omx_utils/src/mfx_omx_defaults.cpp]: Thanks Denis. Already set to 15 for GopPicSize.

File: omx_utils/include/mfx_omx_types.h (hunk 2 above)
Comment (TianmiChen) [reply omx_utils/include/mfx_omx_types.h]: Thanks Denis and Lvan,

This change is used to set some parameters for below structure in the external buffer defined in mfxstructure.h.  Please refer to https://github.com/Intel-Media-SDK/MediaSDK_OMX_IL/blob/b9fa5d1cc04c894585c957f4f2c42e68e363281b/omx_components/src/mfx_omx_venc_component.cpp#L426.

typedef struct {
    mfxExtBuffer Header;

    mfxU16  FrameWidth;
    mfxU16  FrameHeight;

    mfxU16  WriteIVFHeaders;        /* tri-state option */

#if (MFX_VERSION >= MFX_VERSION_NEXT)
    mfxI16  LoopFilterRefDelta[4];
    mfxI16  LoopFilterModeDelta[2];
#else // API 1.26
    mfxI16  reserved1[6];
#endif
    mfxI16  QIndexDeltaLumaDC;
    mfxI16  QIndexDeltaChromaAC;
    mfxI16  QIndexDeltaChromaDC;
#if (MFX_VERSION >= 1029)
    mfxU16  NumTileRows;
    mfxU16  NumTileColumns;
    mfxU16  reserved[110];
#else
    mfxU16  reserved[112];
#endif
} mfxExtVP9Param;

**Pull request merged**